File: src/AzureDataFactory.TestingFramework/Models/Repositories/DataFactoryRepositoryFactory.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

using System.Text.Json;
using Azure.ResourceManager.DataFactory;

namespace AzureDataFactory.TestingFramework.Models.Repositories;

public static class DataFactoryRepositoryFactory
{
    public static DataFactoryRepository ParseFromFolder(string folderPath)
    {
        var pipelines = GetDataFactoryEntitiesByFolderPath(Path.Combine(folderPath, "pipeline"), Pipeline.DeserializeDataFactoryPipelineData);
        var linkedServices = GetDataFactoryEntitiesByFolderPath(Path.Combine(folderPath, "linkedService"), DataFactoryLinkedServiceData.DeserializeDataFactoryLinkedServiceData);
        var datasets = GetDataFactoryEntitiesByFolderPath(Path.Combine(folderPath, "dataset"), DataFactoryDatasetData.DeserializeDataFactoryDatasetData);
        var triggers = GetDataFactoryEntitiesByFolderPath(Path.Combine(folderPath, "trigger"), DataFactoryTriggerData.DeserializeDataFactoryTriggerData);

        return new DataFactoryRepository(pipelines, linkedServices, datasets, triggers);
    }

    private static List<TType> GetDataFactoryEntitiesByFolderPath<TType>(string folderPath, Func<JsonElement, TType> deserialize) where TType : class
    {
        if (!Directory.Exists(folderPath))
            return new List<TType>();

        return Directory.GetFiles(folderPath, "*.json")
            .Select(File.ReadAllText)
            .Select(json => JsonSerializer.Deserialize<JsonElement>(json))
            .Select(deserialize)
            .ToList();
    }
}

File: src/python/data_factory_testing_framework/models/repositories/data_factory_repository_factory.py
import json
import os

from data_factory_testing_framework.generated import Deserializer
from data_factory_testing_framework.models.repositories.data_factory_repository import DataFactoryRepository
from data_factory_testing_framework.models.repositories.models_repository import ModelsRepository

models_repository = ModelsRepository()
deserializer = Deserializer(models_repository.get_models())


class DataFactoryRepositoryFactory:

    @staticmethod
    def parse_from_folder(folder_path: str):
        pipelines = DataFactoryRepositoryFactory._get_data_factory_entities_by_folder_path(folder_path, "PipelineResource")
        return DataFactoryRepository(pipelines)

    @staticmethod
    def _get_data_factory_entities_by_folder_path(folder_path: str, target_class: str):
        entities = []
        files = os.listdir(folder_path)
        for file in files:
            if file.endswith(".json"):
                with (open(os.path.join(folder_path, file), 'r') as f):
                    pipeline_json = json.load(f)
                    entities.append(deserializer._deserialize(target_class, pipeline_json))

        return entities
